core/src/KERC20.sol

```diff
@@ -0,0 +1,198 @@
+        unchecked {
+            balanceOf[to] += amount;
+        }
+
+        emit Transfer(from, to, amount);
+
+        return true;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             EIP-2612 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function permit(
+        address owner,
+        address spender,
+        uint256 value,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public virtual {
+        require(deadline >= block.timestamp, "PERMIT_DEADLINE_EXPIRED");
+
+        // Unchecked because the only math done is incrementing
+        // the owner's nonce which cannot realistically overflow.
+        unchecked {
+            address recoveredAddress = ecrecover(
+                keccak256(
+                    abi.encodePacked(
+                        "\x19\x01",
+                        DOMAIN_SEPARATOR(),
+                        keccak256(
+                            abi.encode(
+                                keccak256(
+                                    "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+                                ),
+                                owner,
+                                spender,
+                                value,
+                                nonces[owner]++,
+                                deadline
+                            )
+                        )
+                    )
+                ),
+                v,
+                r,
+                s
+            );
+
+            require(recoveredAddress != address(0) && recoveredAddress == owner, "INVALID_SIGNER");
+
+            allowance[recoveredAddress][spender] = value;
+        }
+
+        emit Approval(owner, spender, value);
+    }
+
+    function DOMAIN_SEPARATOR() public view virtual returns (bytes32) {
+        return block.chainid == INITIAL_CHAIN_ID ? INITIAL_DOMAIN_SEPARATOR : computeDomainSeparator();
+    }
+
+    function computeDomainSeparator() internal view virtual returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"),
+                    keccak256(bytes(name)),
+                    keccak256("1"),
+                    block.chainid,
+                    address(this)
+                )
+            );
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        INTERNAL MINT/BURN LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev You must do `totalSupply += amount` separately. Do so in a checked context.
+    function _unsafeMint(address to, uint256 amount) internal virtual {
+        unchecked {
+            balanceOf[to] += amount;
+        }
```

Comment:
This file was taken from https://github.com/transmissions11/solmate/blob/main/src/tokens/ERC20.sol and lightly modified. The original `_mint` function looked like this:
```Solidity
function _mint(address to, uint256 amount) internal virtual {
    totalSupply += amount;

    // Cannot overflow because the sum of all user
    // balances can't exceed the max uint256 value.
    unchecked {
        balanceOf[to] += amount;
    }

    emit Transfer(address(0), to, amount);
}
```

In this version, `_mint` both reads and writes from whatever storage slot `totalSupply` is in. This is gassy! In my version, `totalSupply` is stored as a `uint112` instead of `uint256`, allowing me to pack other variables into the same storage slot:
- `uint112 totalSupply`
- `uint112 lastBalance`
- `uint32 lastAccrualTime`
Note that these sum to 256 bits, which means I'm using every last bit of the slot.

So, getting back to `_mint`: If we keep the `totalSupply += amount` inside this function, it would copy the entire slot from storage, update the `totalSupply` portion in memory, and then copy the updated one back to storage. It'd be far better to only write to the packed slot once I have **all** the updated info -- `totalSupply, `lastBalance` and `lastAccrualTime`. To make that happen, I remove the `totalSupply` increment from here and do it elsewhere. BUT in order to avoid overflows in the `unchecked` balanceOf block, it's important that regardless of where I increment totalSupply, I do it in a checked context.